assets/model_monitoring/components/src/shared_utilities/io_utils.py

```diff
@@ -14,45 +24,60 @@ def init_spark():
+                      "most likely there is no request in the specified window.")
+    elif "The requested stream was not found" in error.args[0]:
+        return True, (f"No data is found for input '{input_name}', "
+                      "seems the root folder of the input is moved or deleted.")
+    elif "Not able to find MLTable file" in error.args[0]:
+        return True, f"No data is found for input '{input_name}'."
+    else:
+        return False, None
 
 
-def try_read_mltable_in_spark_with_warning(
-    mltable_path: str, input_name: str
-) -> DataFrame:
+def try_read_mltable_in_spark_with_warning(mltable_path: str, input_name: str) -> DataFrame:
     """Read mltable in spark. In case of failure, posts a warning to the job and returns None."""
-    try:
-        return read_mltable_in_spark(mltable_path)
-    except Exception as error:
-        if _is_not_found_exception(error):
-            print(error)
-            error_message = f"No data was found for input '{input_name}'."
-            print(error_message)
+    return try_read_mltable_in_spark(mltable_path, input_name, NoDataApproach.WARNING)
+
+
+def try_read_mltable_in_spark_with_error(mltable_path: str, input_name: str) -> DataFrame:
+    """Read mltable in spark."""
+    return try_read_mltable_in_spark(mltable_path, input_name, NoDataApproach.ERROR)
+
+
+def try_read_mltable_in_spark(mltable_path: str, input_name: str, no_data_approach=NoDataApproach.IGNORE) -> DataFrame:
+    """
+    Read mltable in spark.
+
+    If data not found, conduct different error handling based on no_data_approach.
+    """
+    def process_input_not_found(error_message):
+        if no_data_approach == NoDataApproach.IGNORE:
+            return None
+        elif no_data_approach == NoDataApproach.WARNING:
             post_warning_event(
                 error_message
                 + " Please visit aka.ms/mlmonitoringhelp for more information."
             )
-        else:
-            raise error
-        return None
-
+            return None
+        else:  # no_data_approach == NoDataApproach.ERROR:
+            raise DataNotFoundError(error_message)
 
-def try_read_mltable_in_spark(mltable_path: str, input_name: str) -> DataFrame:
-    """Read mltable in spark. In case of failure, returns None."""
     try:
-        return read_mltable_in_spark(mltable_path)
+        df = read_mltable_in_spark(mltable_path)
     except Exception as error:
-        if _is_not_found_exception(error):
+        input_not_found, error_message = _is_not_found_exception(input_name, error)
+        if input_not_found:
             print(error)
-            error_message = f"No data was found for input '{input_name}'."
-            print(error_message)
+            return process_input_not_found(error_message)
         else:
             raise error
+    if (not df) or df.isEmpty():
+        error_message = (f"No data is found for input '{input_name}' in the specified window, "
```

Comment:
can we also print out the time window string

Reply:
time window string is applied in _raw_mdc_uri_folder_to_mltable(), after that, only mltable is passed down, all the detail including time window is in the MLTable. Added mltable path to the error msg, user can check it by themselves.